orion-lib/src/listeners/listeners_manager.rs

```diff
@@ -125,35 +159,76 @@ impl ListenersManager {
     }
 
     pub fn start_listener(&mut self, listener: Listener, listener_conf: ListenerConfig) -> Result<()> {
-        let listener_name = listener.get_name();
+        let listener_name = listener.get_name().to_string();
         let (addr, dev) = listener.get_socket();
         info!("Listener {} at {addr} (device bind:{})", listener_name, dev.is_some());
-        // spawn the task for this listener address, this will spawn additional task per connection
+
+        self.version_counter += 1;
+        let version = self.version_counter;
+
+        let listener_name_for_async = listener_name.clone();
         let join_handle = tokio::spawn(async move {
             let error = listener.start().await;
-            warn!("Listener {listener_name} exited: {error}");
+            info!("Listener {} version {} exited: {}", listener_name_for_async, version, error);
         });
-        #[cfg(debug_assertions)]
-        if self.listener_handles.contains_key(&listener_name) {
-            debug!("Listener {listener_name} already exists, replacing it");
-        }
-        // note: join handle gets overwritten here if it already exists.
-        // handles are abort on drop so will be aborted, closing the socket
-        // but the any tasks spawned within this task, which happens on a per-connection basis,
-        // will survive past this point and only get dropped when their session ends
-        self.listener_handles.insert(listener_name, ListenerInfo::new(join_handle, listener_conf));
+
+        let listener_info = ListenerInfo::new(join_handle, listener_conf, version);
+
+        self.listener_handles.insert(listener_name.clone(), listener_info);
+        let version_count = self.listener_handles.get_vec(&listener_name).map(|v| v.len()).unwrap_or(0);
+        info!("Started version {} of listener {} ({} total active version(s))", version, listener_name, version_count);
+
+        self.cleanup_old_versions(&listener_name);
 
         Ok(())
     }
 
     pub fn stop_listener(&mut self, listener_name: &str) -> Result<()> {
-        if let Some(abort_handler) = self.listener_handles.remove(listener_name) {
-            info!("{listener_name} : Stopped");
-            abort_handler.handle.abort();
+        if let Some(listeners) = self.listener_handles.remove(listener_name) {
+            info!("Stopping all {} version(s) of listener {}", listeners.len(), listener_name);
+            for listener_info in listeners {
+                info!("Stopping listener {} version {}", listener_name, listener_info.version);
+                listener_info.handle.abort();
+            }
+        } else {
+            info!("No listeners found with name {}", listener_name);
         }
 
         Ok(())
     }
+
+    fn cleanup_old_versions(&mut self, listener_name: &str) {
+        if let Some(mut versions) = self.listener_handles.remove(listener_name) {
```

Comment:
I think you could use ```self.listener_handles.get_vec_mut``` instead of ```remove``` so then you won't need to re-add listeners in 217 and you won't need to make a call at 22.

Because of line 206, there is no risk of draining all listeners.